libeos-updater-util/flatpak.c

```diff
@@ -0,0 +1,1411 @@
+  const gchar *collection_id = NULL;
+  FlatpakRefKind kind;
+
+  if (!parse_flatpak_ref_from_entry (entry, &app_name, &kind, error))
+    return NULL;
+
+  collection_id = json_object_get_string_member (entry, "collection-id");
+
+  if (collection_id == NULL)
+    {
+      g_set_error (error,
+                   EOS_UPDATER_ERROR,
+                   EOS_UPDATER_ERROR_MALFORMED_AUTOINSTALL_SPEC,
+                   "Expected a 'collection-id' member");
+      return NULL;
+    }
+
+  /* TODO: Right now we "stuff" the collection-id in the remote-name part
+   * of the FlatpakRemoteRef and look up the corresponding remote later on
+   * when actually pulling the flatpaks */
+  return g_object_new (FLATPAK_TYPE_REMOTE_REF,
+                       "remote-name", collection_id,
+                       "name", app_name,
+                       "kind", kind,
+                       NULL);
+}
+
+static FlatpakRemoteRef *
+flatpak_remote_ref_from_uninstall_action_entry (JsonObject  *entry,
+                                                GError     **error)
+{
+  const gchar *app_name = NULL;
+  FlatpakRefKind kind;
+
+  if (!parse_flatpak_ref_from_entry (entry, &app_name, &kind, error))
+    return NULL;
+
+  return g_object_new (FLATPAK_TYPE_REMOTE_REF,
+                       "remote-name", "none",
+                       "name", app_name,
+                       "kind", kind,
+                       NULL);
+}
+
+static FlatpakRemoteRef *
+flatpak_remote_ref_from_action_entry (EosUpdaterUtilFlatpakRemoteRefActionType   action_type,
+                                      JsonObject                                *entry,
+                                      GError                                   **error)
+{
+  switch (action_type)
+    {
+      case EUU_FLATPAK_REMOTE_REF_ACTION_INSTALL:
+        return flatpak_remote_ref_from_install_action_entry (entry, error);
+      case EUU_FLATPAK_REMOTE_REF_ACTION_UNINSTALL:
+        return flatpak_remote_ref_from_uninstall_action_entry (entry, error);
+      default:
+        g_assert_not_reached ();
+    }
+}
+
+static FlatpakRemoteRefAction *
+flatpak_remote_ref_action_from_json_node (JsonNode *node,
+                                          GError **error)
+{
+  const gchar *action_type_str = NULL;
+  JsonObject *object = json_node_get_object (node);
+  g_autoptr(FlatpakRemoteRef) flatpak_remote_ref = NULL;
+  g_autoptr(GError) local_error = NULL;
+  JsonNode *serial_node = NULL;
+  gint64 serial64;
+  gint32 serial;
+  EosUpdaterUtilFlatpakRemoteRefActionType action_type;
+
+  if (object == NULL)
+    {
+      g_autofree gchar *node_str = json_node_to_string (node);
+      g_set_error (error,
+                   EOS_UPDATER_ERROR,
+                   EOS_UPDATER_ERROR_MALFORMED_AUTOINSTALL_SPEC,
+                   "Expected entry node to be an object, but was %s", node_str);
+
+      return NULL;
+    }
+
+  action_type_str = json_object_get_string_member (object, "action");
```

Comment:
Unfortunately the `json_object_get_*_member()` methods will hit an assertion failure if no member exists by that name, or if it has the wrong type. You need to use `json_object_get_member()`, check it’s non-`NULL`, check its type, then get its string value. :-(

Reply:
That's unfortunate. Would a patch to json-glib for `json_object_safe_get_*_member` be accepted?

Reply:
I would *love* to see that. @ebassi is the person to speak to, since he maintains json-glib.

Reply:
See [this issue](https://gitlab.gnome.org/GNOME/json-glib/issues/4).

Short answer: no, not really.

Longer answer: JSON has no validating schema, and C does not have an "unset" value; a missing object member returning `NULL` or `0` is indistinguishable from an existing object member holding that value.

The usual pattern for GLib-based API would be to have a `json_object_get_<type>_member()` version with a `GError` argument, and let you handle the missing key error condition; I find it distasteful in `GKeyFile` not only because it makes for heavily convoluted error handling code, but also because it will require the returned value to be an out parameter, in order to be handled correctly by language bindings. So I'm not going to accept a patch that does that.

The only suitable replacement would be a Python-esque `json_object_get_<type>_member_with_default()` that allows you to specify the default value in case the member is missing. The API would be quite a mouthful, but I'd consider it the only feasible approach.

Reply:
OSTree uses out-parameters to return values quite successfully. I’d be happy with either of:
```
gboolean json_object_get_<type>_member (JsonObject *obj, const gchar *member_name, <type> *out_value);
```
which returns `TRUE` iff the member exists and is of `<type>` (and hence has been returned in `out_value`).

*Or*

```
<type> json_object_get_<type>_member_with_default (JsonObject *obj, const gchar *member_name, const <type> default);
```